File: Factoring/prime_factorization.cpp
#include <iostream>
#include <vector>
#include <math.h>

using namespace std;

vector<int> pcSplit(int i)
{
    int p = 2;
    while(i % p != 0)
    {
        p++;
    }

    vector<int> ret = {p, i / p};
    return ret;
}

bool isPrime(int j)
{
    bool prime = true;

    for (int div = 2; div <= pow(j, 0.5); div++)
    {
        if (j % div == 0)
        {
            prime = false;
        }
    }

    return prime;
}

vector<int> primeFactors(int n)
{
    vector<int> ends;

    if (n <= 1)
    {
        return ends;
    }

    else
    {
    while (true)
        {
            vector<int> split = pcSplit(n);
            ends.push_back(split[0]);
            n = split[1];

            if (isPrime(split[0]) && isPrime(split[1]))
            {
                ends.push_back(split[1]);
                break;
            }
        }
    }


    return ends;
}

string printVector(vector<int> vec)
{
    for(int i = 0; i < vec.size(); ++i)
        cout << vec[i] << ' ';
}

int main()
{
    int toFactor;

    cout << "Number to factor: ";
    cin >> toFactor;

    vector<int> primeFactorization = primeFactors(toFactor);
    printVector(primeFactorization);
}

File: Factoring/prime_factorization.py
from time import time


def prime_factors(n):
    ends = []

    def pc_split(i):
        p = 2
        while i % p != 0:
            p += 1

        return p, i // p

    def is_prime(j):
        prime = True
        div = 2

        while div <= j ** (1 / 2):
            if j % div == 0:
                prime = False

            div += 1

        return prime

    if n <= 1:
        return None

    else:
        while 1:
            split = pc_split(n)
            n = split[1]
            ends.append(split[0])

            if is_prime(split[0]) and is_prime(split[1]):
                ends.append(split[1])
                break

        return ends


t1 = time()
print(prime_factors(-2))
t2 = time()

print(f"Factorization took {t2 - t1} secs")
